src/sdk/namespace_fs.js

```diff
@@ -877,6 +918,55 @@ class NamespaceFS {
         }
     }
 
+    // Allocated config.NSFS_BUF_SIZE in Semaphore but in fact we can take up more inside
+    // This is due to MD5 calculation and data buffers
+    // Can be finetuned further on if needed and inserting the Semaphore logic inside
+    // Instead of wrapping the whole _upload_stream function (q_buffers lives outside of the data scope of the stream)
+    async _upload_stream_tmp(params, upload_path, file_path, fs_account_config, rpc_client, fs_xattr) {
+        let target_file;
+        const { source_stream, md5_b64, key, bucket, upload_id } = params;
+        try {
+            target_file = await nb_native().fs.open(fs_account_config, upload_path, 'wt', get_umasked_mode(config.BASE_MODE_FILE));
+            // Not using async iterators with ReadableStreams due to unsettled promises issues on abort/destroy
+            const chunk_fs = new ChunkFS({
+                target_file,
+                fs_account_config,
+                rpc_client,
+                namespace_resource_id: this.namespace_resource_id
+            });
+            chunk_fs.on('error', err1 => dbg.error('namespace_fs._upload_stream: error occured on stream ChunkFS: ', err1));
+            await stream_utils.pipeline([source_stream, chunk_fs]);
+            await stream_utils.wait_finished(chunk_fs);
+            if (chunk_fs.digest) {
+                if (md5_b64) {
+                    const md5_hex = Buffer.from(md5_b64, 'base64').toString('hex');
+                    if (md5_hex !== chunk_fs.digest) throw new Error('_upload_stream mismatch etag: ' + util.inspect({ key, bucket, upload_id, md5_hex, digest: chunk_fs.digest }));
+                }
+                fs_xattr = this._assign_md5_to_fs_xattr(chunk_fs.digest, fs_xattr);
+            }
+            if (fs_xattr) {
+                await target_file.setxattr(fs_account_config, fs_xattr);
+            }
+            if (config.NSFS_TRIGGER_FSYNC) await target_file.fsync(fs_account_config);
+            // Used for etag
+            return fs_xattr;
+        } catch (error) {
+            dbg.error('_upload_stream had error: ', error);
+            throw error;
+        } finally {
+            try {
+            dbg.log1('NamespaceFS.link_object:', target_file, upload_path);
+            if (typeof nb_native().fs.temp_file === 'function') {
+                await target_file.linkfileat(fs_account_config, upload_path, file_path);
+            }
```

Comment:
I think this should not be in the catch, but instead moved up into the try block, so that we will still close the file on failure to link.

Reply:
right, it's inside a lower try block with the call to `close`, which means that it won't close if linkAt fails. But there's another try surrounding it, and there we are inside a finally block actually. So Let me resolve this thread, because the code moved with the merge of the functions. I will suggest what to move on the updated code in a moment.